### PR TITLE
feat(dataset): import dataset at specific tags

### LINKED ISSUE #2904: Allow importing datasets at a specific tag
Renku allows users to version their datasets through a tags mechanism.
To improve the UX around dataset versioning, users should be able to specify a tag when importing datasets into their Renku project. For example through a `renku dataset import --tag <tag>` option.

### LINKED ISSUE #2910: renku dataset unlink not working for imported datasets
**Describe the bug**
`renku dataset unlink` correctly unlinks files in the home project of the dataset, however when the dataset was imported in other projects, `renku dataset update` does not remove the files.

**Link to project**

In the dataset of [this project](https://renkulab.io/projects/cyril.matthey-doret/statbot-parquet2postgres/overview/commits), I replaced file `metadata/1_01_001_CH.yml` by `metadata/1_01_001_BS.yml` (commits d9278d7e and 9506d4dd).

In another project where [the dataset is imported](https://renkulab.io/projects/cyril.matthey-doret/statbot-db-migration/datasets/statbot-sample/), I ran `renku dataset update statbot-sample`. The new file `metadata/1_01_001_BS.yml` was added, but `metadata/1_01_001_CH.yml` was not unlinked from the dataset. 

**To Reproduce**
1. Create a dataset
2. import the dataset in a separate project
3. Unlink files in the parent project
4. Update the imported dataset

**Expected behavior**
Both datasets are kept in sync and contain the same files

**Run environment (please complete the following information):**
 - renku-python 1.2.2
 - Python 3.9.7
 - local terminal

### PR DESCRIPTION
**Pull request by mohammad-sdsc**
# Description

Allows specifying a dataset tag when importing a Renku dataset. It then imports the specified version (files and metadata) instead of the latest version.

Fixes #2904 
Fixes #2910

### CODE REVIEW

File: renku/core/dataset/providers/renku.py [hunk 1]
```diff
@@ -388,10 +367,26 @@ def _fetch_dataset(self, client_dispatcher: IClientDispatcher, database_dispatch
...
+                if tag is None:
+                    raise errors.ParameterError(f"Cannot find tag '{self._tag}' for dataset '{self._name}'")
+
+                dataset = datasets_provenance.get_by_id(tag.dataset_id.value)
+
+            assert dataset is not None
             provider_dataset = ProviderDataset.from_dataset(dataset)
+
+            # NOTE: Set the dataset version so that it can be checked later to see if a tag was specified for import
+            provider_dataset.version = self._tag
```
Comment (Panaetius) [inline renku/core/dataset/providers/renku.py:389]: There is code on import to create a tag if `.version` is set. Would this result in duplicate tags being created?

**Review by Panaetius — COMMENTED** [review]
Thank you for removing attrs!

I think the whole dataset add/import logic has grown a bit organically over the years, and it's become a bit hard to follow in parts, with many edge cases and different conditions. I wonder if at some point we should refactor it, with dedicated classes for different behaviors. Similar to how we reworked the template logic.
That might make it easier to understand and maintain.

File: renku/ui/cli/dataset.py [hunk 2]
```diff
@@ -934,12 +949,44 @@ def export_(name, provider, publish, tag, **kwargs):
...
+            for j, (param_name, (param_description, param_type)) in enumerate(params.items()):
+                if j == 0:
+                    param_description = f"\b\n{param_description}\n "  # NOTE: add newline after a group
+                f = optgroup.option(f"--{param_name}", type=param_type, help=param_description, default=None)(f)
+
+            name = f"{name} configuration"
+            if i == len(providers) - 1:
+                name = "\n  " + name  # NOTE: add newline before first group
+
+            f = optgroup.group(name=name)(f)
```
Comment (Panaetius) [inline renku/ui/cli/dataset.py:976]: Probably something for a separate issue, but it'd be really nice if we could return a `Proxy` here, for import times. If each proxy returns a wrapped Proxy unless there's no `next()` that should work I think, and only be evaluated if this method is actually accessed. I.e. Proxy wrapping the whole decorator instead of just wrapping a value inside the decorator. But I'm not sure if click accepts that

File: renku/ui/service/serializers/datasets.py [hunk 3]
```diff
@@ -181,6 +181,7 @@ class DatasetImportRequest(AsyncSchema, LocalRepositorySchema, RemoteRepositoryS
     dataset_uri = fields.String(required=True)
     name = fields.String(description="Optional dataset name.")
     extract = fields.Boolean()
+    tag = fields.String(description="Dataset version to import.")
```
Comment (Panaetius) [inline renku/ui/service/serializers/datasets.py:184]: We need to bump the API (minor) version when adding fields, so a client can know whether this feature is supported or not.

See https://github.com/SwissDataScienceCenter/renku-python/blob/develop/renku/ui/service/views/api_versions.py#L62 and add the version here: https://github.com/SwissDataScienceCenter/renku-python/blob/develop/renku/ui/service/views/datasets.py#L191 (And all other endpoints that have 1.1, which is annoying)

File: renku/core/dataset/providers/olos.py [hunk 4]
```diff
@@ -81,27 +80,25 @@ def set_parameters(self, client_dispatcher: IClientDispatcher, *, dlcm_server=No
...
+        self.server_url = server_url
 
     def set_access_token(self, access_token):
         """Set access token."""
         self.access_token = access_token
 
     def access_token_url(self):
         """Endpoint for creation of access token."""
-        return urllib.parse.urljoin(self._server_url, "/portal by clicking on the top-right menu and selecting 'token'")
+        return urllib.parse.urljoin(self.server_url, "/portal by clicking on the top-right menu and selecting 'token'")
```
[the remark this answers is not in the document]
Comment (mohammad-sdsc) [reply renku/core/dataset/providers/olos.py:97]: Hmmmmm, who might have written it 😄. I'll remove the text after since it might change without us knowing.

File: renku/infrastructure/repository.py [hunk 5]
```diff
@@ -198,6 +184,23 @@ def is_valid(self) -> bool:
...
+            command += ["--skip-smudge"]
+
+        try:
+            self.run_git_command(*command)
+        except errors.GitCommandError as e:
+            raise errors.GitError(f"Cannot install Git LFS in {self}") from e
+
+    def get_changes_patch(self, path: Union[Path, str]) -> List[str]:
+        """Return a patch of all changes to a file."""
+        return self.run_git_command("log", "--", str(path), patch=True, follow=True).splitlines()
```
[the remark this answers is not in the document]
Comment (mohammad-sdsc) [reply renku/infrastructure/repository.py:202]: I've renamed it to `get_historical_changes_patch`

File: renku/core/dataset/providers/renku.py (hunk 1 above)
Comment (Panaetius) [reply renku/core/dataset/providers/renku.py:389]: But it has a different description, "Tag generated by import"

File: renku/core/dataset/providers/olos.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/dataset/providers/olos.py:97]: It's annoying that `git blame` doesn't work across moves/renames. I'm responsible for so many bugs now :see_no_evil:

[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/dataset/providers/olos.py:97]: :cry: I blame OLOS

File: renku/core/dataset/providers/renku.py (hunk 1 above)
Comment (mohammad-sdsc) [reply renku/core/dataset/providers/renku.py:389]: I'm not sure what's the right thing to do here: On one hand we lose the description of the original tag (if there is one) and on the other hand we lose the description that says this tag was imported. I went with this approach since we have the original tag name. I can change it to keep the original tag as well. What do you think?

File: renku/infrastructure/repository.py [hunk 6]
```diff
@@ -93,7 +78,8 @@ def __exit__(self, *args):
         self.close()
 
     def __del__(self):
-        self.close()
+        # self.close()
```
[the remark this answers is not in the document]
Comment (mohammad-sdsc) [reply renku/infrastructure/repository.py:81]: We call `close` on a `Repo` object and then delete it: https://github.com/SwissDataScienceCenter/renku-python/blob/31f9514c19f97a1260b763cf752326d685fdf2b0/renku/infrastructure/repository.py#L694.

But `Repo.__del__` calls `Repo.close` underneath, so, calling close is not needed and removing it resolves the issue.

[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/infrastructure/repository.py]: from what I understand, deleting does not necessarily call __del__, but we should delete it as well so its reference is gone

From the docs:
image
#justpythonthings

**Review by Panaetius — APPROVED** [review]
Thank you

**Review by Panaetius — APPROVED** [review]

**Review by Panaetius — APPROVED** [review]
